vowpalwabbit/array_parameters.h

```diff
@@ -22,153 +22,170 @@ typedef std::unordered_map<uint64_t, weight*> weight_map;
-	typedef  T* pointer;
-	typedef  T& reference;
-
-	sparse_iterator(weight_map::iterator& iter, uint32_t stride)
-		: _iter(iter), _stride(stride)
-	{ }
-
-	sparse_iterator& operator=(const sparse_iterator& other)
-	{
-		_iter = other._iter;
-		_stride = other._stride;
-		return *this;
-
-	}
-	uint64_t index() { return _iter->first; }
-
-	T& operator*() { return *(_iter->second); }
-
-	sparse_iterator& operator++()
-	{
-		_iter++;
-		return *this;
-	}
-
-	bool operator==(const sparse_iterator& rhs) const { return _iter == rhs._iter; }
-	bool operator!=(const sparse_iterator& rhs) const { return _iter != rhs._iter; }
-};
+ private:
+  weight_map::iterator _iter;
+  uint32_t _stride;
+
+ public:
+  typedef std::forward_iterator_tag iterator_category;
+  typedef T value_type;
+  typedef ptrdiff_t difference_type;
+  typedef T* pointer;
+  typedef T& reference;
+
+  sparse_iterator(weight_map::iterator& iter, uint32_t stride) : _iter(iter), _stride(stride) {}
+  sparse_iterator& operator=(const sparse_iterator& other)
+  {
+    _iter = other._iter;
+    _stride = other._stride;
+    return *this;
+  }
+  uint64_t index() { return _iter->first; }
+  T& operator*() { return *(_iter->second); }
+  sparse_iterator& operator++()
+  {
+    _iter++;
+    return *this;
+  }
 
+  bool operator==(const sparse_iterator& rhs) const { return _iter == rhs._iter; }
+  bool operator!=(const sparse_iterator& rhs) const { return _iter != rhs._iter; }
+};
 
 class sparse_parameters
 {
-private:
-	weight_map _map;
-	uint64_t _weight_mask;  // (stride*(1 << num_bits) -1)
-	uint32_t _stride_shift;
-	bool _seeded; // whether the instance is sharing model state with others
-	bool _delete;
-	void* default_data;
-  float* default_value;
-public:
-	typedef sparse_iterator<weight> iterator;
-	typedef sparse_iterator<const weight> const_iterator;
  private:
-	void(*fun)(const weight*, void*);
- public:
+  weight_map _map;
+  uint64_t _weight_mask;  // (stride*(1 << num_bits) -1)
+  uint32_t _stride_shift;
+  bool _seeded;  // whether the instance is sharing model state with others
+  bool _delete;
+  void* default_data;
+  float* default_value;
 
-	sparse_parameters(size_t length, uint32_t stride_shift = 0)
-		: _map(),
-		_weight_mask((length << stride_shift) - 1),
-		_stride_shift(stride_shift),
-		_seeded(false), _delete(false), default_data(nullptr),
-    fun(nullptr)
-	{ default_value = calloc_mergable_or_throw<weight>(stride());}
+ public:
+  typedef sparse_iterator<weight> iterator;
+  typedef sparse_iterator<const weight> const_iterator;
 
-	sparse_parameters()
-		: _map(), _weight_mask(0), _stride_shift(0), _seeded(false), _delete(false), default_data(nullptr), fun(nullptr)
-	{ default_value = calloc_mergable_or_throw<weight>(stride());}
+ private:
+  void (*fun)(const weight*, void*);
 
-	bool not_null() { return (_weight_mask > 0 && !_map.empty()); }
+ public:
+  sparse_parameters(size_t length, uint32_t stride_shift = 0)
+      : _map(),
```

Comment:
Can this take fewer lines?

Reply:
It it doesn’t all fit on one line then it splits across multiple. I’d prefer this to be one per line (as is) rather than bin packed

Reply:
is bin packed an option?  The same thing is happening to option declarations which is a hefty chunk of the extra 2k lines of code.

Reply:
I think the fact this is split across lines is good, much easier to read. The option declarations also are very long lines, so it is kind of nice this is splitting this up. The increase in number of lines is not a bad thing in my opinion as it improves overall readability.

Reply:
@palmerlao we want to have the overflow to fill the entire line before it continues onto another, is this possible?